design-proposals/migration-policies.md

```diff
@@ -0,0 +1,227 @@
+the relevant label.
+
+## User Stories
+As an admin I would like to have different migration policies for different groups of VMs.
+* By "migration policies" I mean controlling one or more of the following migration aspects:
+    * Bandwidth
+    * Scheduling
+    * Storage
+* By “different groups of VMs” I could mean:
+    * All VMs that have certain label defined
+    * All VMs that have certain label with certain value defined
+    * All VMs that belong to a namespace with certain label key/value
+* As an admin I know that some workloads are in higher priority than others and therefore need different 
+  migration policies.
+* As an admin I know that some workloads are not migration-friendly and demand careful configurations in order
+  for migrations to converge.
+* As an admin I want to have the ability to temporarily tune migrations for a specific event such as draining
+  a node or performing an upgrade.
+* As an admin I want that the migration policy would be as transparent as possible for the VM owner / creator.
+* As a user I want to be able to tell admins the workload characteristics of my VMs so that admins can create
+  policies and/or namespaces with the desired characteristics.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+The design will be similar to Kubernetes' [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
+which is a similar concept. Therefore, a new MigrationPolicy CRD will be introduced. This CRD's spec will contain all
+relevant fields from KubevirtCR's [MigrationConfiguration](https://kubevirt.io/api-reference/master/definitions.html#_v1_migrationconfiguration)
+(i.e. which are not inherently cluster-wide configurations). In the future, additional
+configurations (or status related fields) that inherently relate to non-global context could be added,
+thus the configurations on NetworkPolicy and KubevirtCR's MigrationConfigurations may diverge to some degree.
+
+In addition, the new CRD's spec will include the following ways of specifying the groups of VMIs on which
+to apply the policy:
+* By VMI's labels
+* By namespace's labels
+
+All of these methods can be combined, for example a policy can require both VMI labels and namespace labels in
+order to match.
+
+## Policies' Precedence
+
+It is possible that a multiple policies apply to the same VMI. In such cases, the precedence is in the
+same order as the bullets above (VMI labels first, then namespace labels). It is not allowed to define
+two policies with the exact same selectors.
+
+If two policies apply to the same VMI, both have VMI:
+* First it is matched by VMI labels, only then to namespace labels
+* If multiple policies match to VMI labels, the policy with the highest number of matched labels has higher precedence.
+* If multiple policies match to VMI labels and have the same number of matched labels precedence is by the matched 
+  labels' lexicographic order.
```

Comment:
shouldn't the most detailed match take precedence? 

So for example, a migration policy that matches a VMI's and the Namespace's labels seems like it is more detailed that a policy that only matches the VMI's labels

Reply:
Do you mean that precedence will be determined by the number of matching labels (whether their namespace or vmi labels)? It makes sense. To me, intuitively, it makes sense that VMI labels will have more effect than namespace labels since they can "override" their namespace default configurations.

For example: as a system admin I imagine defining policies for different namespaces (or namespaces groups) by referring to their labels. If I want to override policies to special VMs regardless of their namespace membership I can define another policy by referring to the VMIs labels.

I'm sure that you're more familiar than I in which use case makes more sense.

Another question that arises: what about a situation in which the same number of labels are matched for two policies that refer to different label types? E.g. two policies are matched, one with 5 VMI labels the other with 5 namespace labels, which policy has precedence?

Reply:
> Another question that arises: what about a situation in which the same number of labels are matched for two policies that refer to different label types? E.g. two policies are matched, one with 5 VMI labels the other with 5 namespace labels, which policy has precedence?

I think we just need to be consistent and select the same policy every we are migrating a VMI. IIRC @davidvossel once suggested sorting the selected policies and always pickup the first one.

Reply:
Sounds good.
Sort by what? Lexicographic order of label keys sounds good? (just for consistency).

Reply:
> Another question that arises: what about a situation in which the same number of labels are matched for two policies that refer to different label types? E.g. two policies are matched, one with 5 VMI labels the other with 5 namespace labels, which policy has precedence?

doesn't matter, just make sure the same one is picked. Lexicographic order by migration policy name from the set of migration policies with the same detailed match.